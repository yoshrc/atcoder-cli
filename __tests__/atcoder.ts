jest.useFakeTimers();
import {username, password} from "./auth.json";
import inquirer from "inquirer";
import {AtCoder} from "../src/atcoder";
import {Cookie} from "../src/cookie";

jest.mock("inquirer");

/*
 * Cookieファイルの実装を差し替えておく
 * コンフィグファイルとの読み書きを行わず、必ず空のセッションが返るようにする
 */
Cookie.prototype.loadConfigFile = jest.fn(async function () {
	// @ts-ignore
	this.cookies = []
});
Cookie.prototype.saveConfigFile = jest.fn(async function () {
});

/*
 このテストが失敗する場合、まず以下の点を確認してください
 - __tests__/auth.json が存在し、正しいユーザー名とパスワードが記述されていること
   すべてのテストを開始する前に、__tests__/auth.example.jsonをコピーして__tests__/auth.json を作成し、
   有効なAtCoderアカウントのユーザー名とパスワードを記述してください
   __tests__/auth.jsonはgit管理に含めないように注意してください
 */
test("AtCoder Login", async () => {
	const atcoder = new AtCoder();
	expect(await atcoder.checkSession()).toBe(false);

	// ユーザー名とパスワードを標準入力で受け付けるかわりにファイルから流し込むようモックする
	// @ts-ignore
	inquirer.prompt.mockResolvedValueOnce({username, password});

	expect(await atcoder.login()).toBe(true);
	expect(await atcoder.checkSession(true)).toBe(true);
});

describe("AtCoder", async () => {
	let atcoder: AtCoder;
	beforeEach(async () => {
		atcoder = new AtCoder();
		// @ts-ignore
		inquirer.prompt.mockResolvedValueOnce({username, password});
		await atcoder.login();
	});
	const contests = ["abc101", "arc101"];
	describe("contest", async () => {
		test.each(contests)("%s", async (contest_id) => {
			expect(await atcoder.contest(contest_id)).toMatchSnapshot();
		});
		test("invalid contest id", async()=>{
			await expect(atcoder.contest("abc0xx")).rejects.toThrow();
		});
	});
	describe("tasks", async () => {
		test.each(contests)("%s", async (contest_id) => {
			expect(await atcoder.tasks(contest_id)).toMatchSnapshot();
		});
		test("invalid contest id", async()=>{
			await expect(atcoder.tasks("abc0xx")).rejects.toThrow();
		});
	});
	const tasks = [["abc101", "abc101_a"], ["abc101", "abc101_b"], ["arc101", "arc101_a"]];
	describe("task", async () => {
		test.each(tasks)("%s %s", async (contest_id, task_id) => {
			expect(await atcoder.task(contest_id, task_id)).toMatchSnapshot();
		});
		test("invalid contest id", async()=>{
			await expect(atcoder.task("abc0xx", "abc0xx_z")).rejects.toThrow();
		});
		test("invalid task id", async()=>{
			await expect(atcoder.task("abc101", "abc102_a")).rejects.toThrow();
		});
	});
});
