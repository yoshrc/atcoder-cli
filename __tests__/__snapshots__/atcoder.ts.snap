// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtCoder contest abc101 1`] = `
Object {
  "id": "abc101",
  "title": "AtCoder Beginner Contest 101",
  "url": "https://atcoder.jp/contests/abc101",
}
`;

exports[`AtCoder contest arc101 1`] = `
Object {
  "id": "arc101",
  "title": "AtCoder Regular Contest 101",
  "url": "https://atcoder.jp/contests/arc101",
}
`;

exports[`AtCoder task abc101 abc101_a 1`] = `
Object {
  "id": "abc101_a",
  "label": "A",
  "title": "Eating Symbols Easy",
  "url": "https://atcoder.jp/contests/abc101/tasks/abc101_a",
}
`;

exports[`AtCoder task abc101 abc101_b 1`] = `
Object {
  "id": "abc101_b",
  "label": "B",
  "title": "Digit Sums",
  "url": "https://atcoder.jp/contests/abc101/tasks/abc101_b",
}
`;

exports[`AtCoder task arc101 arc101_a 1`] = `
Object {
  "id": "arc101_a",
  "label": "C",
  "title": "Candles",
  "url": "https://atcoder.jp/contests/arc101/tasks/arc101_a",
}
`;

exports[`AtCoder tasks abc101 1`] = `
Array [
  Object {
    "id": "abc101_a",
    "label": "A",
    "title": "Eating Symbols Easy",
    "url": "https://atcoder.jp/contests/abc101/tasks/abc101_a",
  },
  Object {
    "id": "abc101_b",
    "label": "B",
    "title": "Digit Sums",
    "url": "https://atcoder.jp/contests/abc101/tasks/abc101_b",
  },
  Object {
    "id": "arc099_a",
    "label": "C",
    "title": "Minimization",
    "url": "https://atcoder.jp/contests/abc101/tasks/arc099_a",
  },
  Object {
    "id": "arc099_b",
    "label": "D",
    "title": "Snuke Numbers",
    "url": "https://atcoder.jp/contests/abc101/tasks/arc099_b",
  },
]
`;

exports[`AtCoder tasks arc101 1`] = `
Array [
  Object {
    "id": "arc101_a",
    "label": "C",
    "title": "Candles",
    "url": "https://atcoder.jp/contests/arc101/tasks/arc101_a",
  },
  Object {
    "id": "arc101_b",
    "label": "D",
    "title": "Median of Medians",
    "url": "https://atcoder.jp/contests/arc101/tasks/arc101_b",
  },
  Object {
    "id": "arc101_c",
    "label": "E",
    "title": "Ribbons on Tree",
    "url": "https://atcoder.jp/contests/arc101/tasks/arc101_c",
  },
  Object {
    "id": "arc101_d",
    "label": "F",
    "title": "Robots and Exits",
    "url": "https://atcoder.jp/contests/arc101/tasks/arc101_d",
  },
]
`;
